bin/runtime/src/migrations/custom_contracts_migration.rs

```diff
@@ -0,0 +1,306 @@
+        let mut num_old_contracts_migrated: u32 = 0;
+        let mut num_old_contracts_skipped: u32 = 0;
+        let mut num_new_contracts: u32 = 0;
+        let mut num_contracts_erred: u32 = 0;
+
+        // Note that we use V = ContractInfo<T> and not V = OldContractInfo<T>, but that does not matter for the prefix.
+        let prefix = ContractInfoOf::<T, ContractInfo<T>>::final_prefix();
+        let mut previous_key = prefix.clone().to_vec();
+        // Iterate over all elements of the ContractInfoOf map
+        while let Some(next) =
+            sp_io::storage::next_key(&previous_key).filter(|n| n.starts_with(&prefix))
+        {
+            previous_key = next;
+            let value = sp_io::storage::get(&previous_key)
+                .expect("The loop conditions checks for existence of the key; qed");
+            match decode_to_either_contract_info::<T>(&mut &value[..]) {
+                Ok(EitherContractInto::Old(old)) => {
+                    if total_items >= MAX_NUM_ITEMS_PROCESSED_PER_MIGRATION {
+                        num_old_contracts_skipped.saturating_accrue(1);
+                        continue;
+                    }
+                    let mut storage_bytes = 0u32;
+                    let mut storage_items = 0u32;
+                    let mut child_key = Vec::new();
+                    while let Some(next_child_key) = child::next_key(&old.trie_id, &child_key) {
+                        child_key = next_child_key;
+                        let mut val_out = [];
+                        let len = child::read(&old.trie_id, &child_key, &mut val_out, 0)
+                            .expect("The loop conditions checks for existence of the key; qed");
+                        storage_bytes.saturating_accrue(len);
+                        storage_items.saturating_accrue(1);
+                    }
+
+                    total_bytes.saturating_accrue(storage_bytes);
+                    total_items.saturating_accrue(storage_items);
+
+                    let storage_byte_deposit =
+                        T::DepositPerByte::get().saturating_mul(storage_bytes.into());
+                    let storage_item_deposit =
+                        T::DepositPerItem::get().saturating_mul(storage_items.into());
+                    let storage_base_deposit = old
+                        .storage_deposit
+                        .saturating_sub(storage_byte_deposit)
+                        .saturating_sub(storage_item_deposit);
+
+                    let new_contract_info = ContractInfo {
+                        trie_id: old.trie_id,
+                        code_hash: old.code_hash,
+                        storage_bytes,
+                        storage_items,
+                        storage_byte_deposit,
+                        storage_item_deposit,
+                        storage_base_deposit,
+                    };
+                    unhashed::put::<ContractInfo<T>>(&previous_key, &new_contract_info);
```

Comment:
is that safe? old version had fewer bytes, could we corrupt db by overwriting smth in memory?

Reply:
`previous_key` is a key corresponding to a single entry in the `ContractInfoOf` map. unhashed::put::<ContractInfo<T>> takes the key and a value, serializes the value, and writes the bytes under this key. We do simply "translate" the old format to the new one, and indeed this code is just lower level details of `translate`.